core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -1060,6 +1084,65 @@ public void setSlaveAgentPort(int port) throws IOException {
         launchTcpSlaveAgentListener();
     }
 
+    /**
+     * Returns the enabled agent protocols.
+     *
+     * @return the enabled agent protocols.
+     * @since FIXME
+     */
+    public Set<String> getAgentProtocols() {
+        if (agentProtocols == null) {
+            // idempotent, so don't care if we do this concurrently, should all get same result
+            Set<String> result = new TreeSet<>();
+            Set<String> disabled = new TreeSet<>();
+            for (String p : StringUtils.split(StringUtils.defaultIfBlank(disabledAgentProtocols, ""), ",")) {
+                disabled.add(p.trim());
+            }
+            Set<String> enabled = new TreeSet<>();
+            for (String p : StringUtils.split(StringUtils.defaultIfBlank(enabledAgentProtocols, ""), ",")) {
+                enabled.add(p.trim());
+            }
+            for (AgentProtocol p : AgentProtocol.all()) {
+                String name = p.getName();
+                if (name != null && (p.isRequired()
+                        || (!disabled.contains(name) && (!p.isOptIn() || enabled.contains(name))))) {
+                    result.add(name);
+                }
+            }
+            agentProtocols = result;
+            return result;
+        }
+        return agentProtocols;
+    }
+
+    /**
+     * Sets the enabled agent protocols.
+     *
+     * @param protocols the enabled agent protocols.
+     * @since FIXME
+     */
+    public void setAgentProtocols(Set<String> protocols) {
+        Set<String> disabled = new TreeSet<>();
+        Set<String> enabled = new TreeSet<>();
+        for (AgentProtocol p : AgentProtocol.all()) {
+            String name = p.getName();
+            if (name != null && !p.isRequired()) {
+                if (p.isOptIn()) {
```

Comment:
Opt-in means it is disabled by default rather than enabled by default.

When we bring out JNLPProtocol4 I will start it as opt-in... Then we will do an A/B test whereby 10% of users are opt-out and 90% are opt-in... After a couple of releases with that we can switch the A/B to 50% and finally turn as opt-out permanently.

I'll probably start an A/B test to turn the legacy protocols to opt-in for 10% of users so we can reach the point where the plaintext protocols are disabled by default... But I'll do that in a separate JIRA